src/orange/surf/detail/FerrariSolver.hh

```diff
@@ -0,0 +1,347 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/surf/detail/FerrariSolver.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cmath>
+#include <iostream>
+
+#include "corecel/Types.hh"
+#include "corecel/cont/Array.hh"
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/PolyEvaluator.hh"
+#include "corecel/math/SoftEqual.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/surf/detail/QuadraticSolver.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Find positive, real, non-zero roots for quartic functions using the
+ * Ferrari-Cardano method\citet{polyanin-ferrari-2007,
+ * https://doi.org/10.1201/9781420010510}.
+ *
+ * The quartic equation
+ * \f[
+   a x^4 + b x^3 + c x^2 + d x + e = 0
+ * \f]
+ * has four solutions mathematically, but we only require solutions which are
+ * both real and positive. This equation is also subject to multiple cases of
+ * catastrophic precision-limitation-based error both fundamentally and as a
+ * consequence of the particular algorithm chosen. This solver implements the
+ * Ferrari-Cardano method, which is well-established and simple, but more
+ * prone to numerical error than contemporary methods to be explored such as
+ * Algorithm 1010\citet{orellana-alg1010-2020,
+ * https://doi.org/10.1145/3386241}.
+ *
+ * \return An Intersections array where each item is either a positive valid
+ * intersection or the sentinel result \c no_intersection().
+ */
+class FerrariSolver
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Intersections = Array<real_type, 4>;
+    using Real2 = Array<real_type, 2>;
+    using Real5 = Array<real_type, 5>;
+    //!@}
+
+    // Solve with manually specified surface state
+    static inline CELER_FUNCTION Intersections solve_general(
+        Real5 abcde,
+        SurfaceState on_surface,
+        real_type tolerance = Tolerance<real_type>::sqrt_quadratic());
```

Comment:
I think with the updated constructor (that has a tolerance) this should be a regular member function, not a static function:

```suggestion
    // Solve with manually specified surface state
   inline CELER_FUNCTION Intersections operator()(
        Real5 abcde,
        SurfaceState on_surface);
```
so if you wanted to change the tolerance you'd write:
```c++
FerrariSolver solve{0.123};
solve({a, b, c, d, e}, SurfaceState::off);
```
etc.